graph.c

```diff
@@ -0,0 +1,350 @@
+        abort();
+    }
+
+    if (weight <= 0) {
+        fprintf(stderr, "Weight should be > 0");
+        abort();
+    }
+}
+
+static void graph_add_new_neighbor(GPtrArray *neighbors, int to_vertice, int weight) {
+    Graph_vertice_neighbor* new_neighbor;
+
+    if ((new_neighbor = (Graph_vertice_neighbor *) malloc(sizeof(*new_neighbor))) == NULL) {
+        fprintf(stderr, "memalloc for Graph_vertice_neighbor failed");
+        abort();
+    }
+
+    new_neighbor->index = to_vertice;
+    new_neighbor->weight = weight;
+
+    g_ptr_array_add(neighbors, new_neighbor);
+}
+
+static void graph_add_neighbors_table(Graph_vertice *self) {
+    GPtrArray *ptr_array = g_ptr_array_new_with_free_func(g_free);
+
+    self->neighbors = ptr_array;
+}
+
+static Graph_vertice_neighbor* graph_find_neighbor(GPtrArray *neighbors, int index) {
+    guint i;
+    Graph_vertice_neighbor* self;
+
+    for (i = 0; i < neighbors->len; i++) {
+        self = g_ptr_array_index(neighbors, i);
+        if(self->index == index) {
+            return self;
+        }
+    }
+
+    return NULL;
+}
+
+
+//add a new weighted directed edge to the graph that connects two vertices
+void graph_add_edge(Graph* self, int from_vertice, int to_vertice, int weight) {
+    Graph_vertice* src;
+    Graph_vertice* dest;
+
+    src = graph_get_vertice(self, from_vertice);
+    dest = graph_get_vertice(self, to_vertice);
+
+    graph_edge_function_check_pararameters(self, src, dest, from_vertice, to_vertice, weight);
+
+    if (dest->neighbors != NULL) {
+        if (dest->neighbors->len > 0) {
+            Graph_vertice_neighbor* n;
+
+            if((n = graph_find_neighbor(dest->neighbors, from_vertice)) != NULL) {
+                n = NULL;
+
+                fprintf(stderr, "Reverse path is found in to_vertice");
+                abort();
+            }
+
+            n = NULL;
+        }
+    }
+
+    if (src->neighbors != NULL) {
+        if (src->neighbors->len > 0) {
+            Graph_vertice_neighbor* n;
+
+            if((n = graph_find_neighbor(src->neighbors, to_vertice)) != NULL) {
+                n->weight = weight;
+                n = NULL;
+
+                return;
+            }
+        }
+    }
+    else {
+        graph_add_neighbors_table(src);
+    }
+
+    graph_add_new_neighbor(src->neighbors, to_vertice, weight);
+
+}
+
+/*
+    erase edge from_vertice to_vertice if there is an edge
+    namely delete neighbor of from_vertice vertice
+*/
+bool graph_erase_edge(Graph *self, int from_vertice, int to_vertice) {
+    Graph_vertice* src;
+    Graph_vertice* dest;
+    int stub_weight = 1;
+
+    src = graph_get_vertice(self, from_vertice);
+    dest = graph_get_vertice(self, to_vertice);
+
+    graph_edge_function_check_pararameters(self, src, dest, from_vertice, to_vertice, stub_weight);
+
+
+    if (dest->neighbors != NULL) {
+        if (dest->neighbors->len > 0) {
+            Graph_vertice_neighbor* n;
```

Comment:
what's the point in n?

Reply:
Just for clarity. I delete it

Reply:
I delete it if I find a way to redefine `if` condition

Reply:
Found and deleted